src/connector/src/sink/file_sink/mod.rs

```diff
@@ -0,0 +1,230 @@
+            };
+        }
+
+        Ok(())
+    }
+
+    async fn update_vnode_bitmap(&mut self, _vnode_bitmap: Arc<Bitmap>) -> Result<()> {
+        Ok(())
+    }
+}
+
+impl OpenDalSinkWriter {
+    pub fn new(
+        operator: Operator,
+        write_path: &str,
+        rw_schema: Schema,
+        pk_indices: Vec<usize>,
+        is_append_only: bool,
+        executor_id: u64,
+        encode_type: SinkEncode,
+    ) -> Result<Self> {
+        let arrow_schema = convert_rw_schema_to_arrow_schema(rw_schema)?;
+        Ok(Self {
+            schema: Arc::new(arrow_schema),
+            write_path: write_path.to_string(),
+            pk_indices,
+            operator,
+            sink_writer: None,
+            is_append_only,
+            epoch: None,
+            executor_id,
+            encode_type,
+        })
+    }
+
+    async fn create_object_writer(&mut self, epoch: u64) -> Result<OpendalWriter> {
+        // Todo: specify more file suffixes based on encode_type.
+        let suffix = match self.encode_type {
+            SinkEncode::Json => "json",
+            SinkEncode::Parquet => "parquet",
+            _ => unimplemented!(),
+        };
+
+        // Note: sink decoupling is not currently supported, which means that output files will not be batched across checkpoints.
+        // The current implementation writes files every time a checkpoint arrives, so the naming convention is `epoch + executor_id + .suffix`.
+        let object_name = format!(
+            "{}/epoch_{}_executor_{}.{}",
+            self.write_path, epoch, self.executor_id, suffix,
+        );
+        Ok(self
+            .operator
+            .writer_with(&object_name)
+            .concurrent(8)
+            .buffer(SINK_WRITE_BUFFER_SIZE)
+            .await?)
+    }
+
+    async fn create_sink_writer(&mut self, epoch: u64) -> Result<()> {
+        let object_writer = self.create_object_writer(epoch).await?;
+        match self.encode_type {
+            SinkEncode::Parquet => {
+                let props = WriterProperties::builder();
+                self.sink_writer = Some(FileWriterEnum::ParquetFileWriter(
+                    AsyncArrowWriter::try_new(
+                        object_writer,
+                        self.schema.clone(),
+                        SINK_WRITE_BUFFER_SIZE,
+                        Some(props.build()),
+                    )?,
+                ));
+            }
+            SinkEncode::Json => {
+                self.sink_writer = Some(FileWriterEnum::FileWriter(object_writer));
+                unimplemented!();
+            }
+            _ => unimplemented!(),
+        }
+
+        Ok(())
+    }
+
+    async fn append_only(&mut self, chunk: StreamChunk) -> Result<()> {
+        let (mut chunk, ops) = chunk.compact().into_parts();
+        let filters =
+            chunk.visibility() & ops.iter().map(|op| *op == Op::Insert).collect::<Bitmap>();
+        chunk.set_visibility(filters);
```

Comment:
hmmm, this seems to implictly convert a retractable stream chunk into an append-only stream chunk. Is it possible to receive a stream chunk with `op != Op::Insert` here? If yes, I feel that we should give an error, ask user to set `force_append_only`, and we only apply this filtering when the option is set. If not, we should add an assertion here instead of filter.

Reply:
Currently for file sink, user need to explicitly support `force_append_only=true`.